packages/server/lib/modes/run.ts

```diff
@@ -529,6 +525,33 @@ async function waitForBrowserToConnect (options: { project: Project, socketId: s
       errors.log(err)
 
       onError(err)
+    }
+
+    return Bluebird.all([
+      waitForSocketConnection(project, socketId),
+      // TODO: remove the need to extend options and coerce this type
+      launchBrowser(options as typeof options & { setScreenshotMetadata: SetScreenshotMetadata }),
+    ]).catch((e) => {
+      // if the error wrapped is a CDPFailedToStartFirefox, try to relaunch the browser
+      if (e.details.includes('CDPFailedToStartFirefox')) {
+        // if CDP fails to connect, which is ultimately out of our control and in the hands of webdriver
+        // we retry launching the browser in the hopes the session is spawned correctly
+        debug(`Caught in launchBrowser: ${e.details}`)
+
+        return retryOnError(e)
```

Comment:
Consider using `util/async_retry.ts` - I'd like to limit how many custom retry implementations we have going!

Reply:
I'm honestly hesitant to touch this code further since it's difficult to test and is incredibly sensitive, hence why I more so going with the patch approach. I also don't know if we have a lot to gain either since we still need to define the `asyncRetry` body  to be referenced twice.